src/scenes/pt.h

```diff
@@ -128,16 +137,23 @@ struct alignas(16) Object {
   }
 
   CU_HD Intersection intersect(Ray const &r) const;
+  CU_D ObjectSample sample(Vec3 const &dir, TraceContext &ctx) const;
 };
 
 struct DeviceScene {
   cuda::raw_ptr<Object> objects;
-  CU_HD Intersection intersect(Ray const &r) const {
+  CU_HD Intersection intersect(Ray const &r, float dist = 0) const {
     Intersection ret;
     for (auto &o : objects) {
       auto test = o.intersect(r);
-      if (test < ret)
+      if (test < ret) {
         ret = test;
+        if (dist) {
+          if (ret.distance + EPSILON < dist) {
```

Comment:
I think at this point ret could still be invalid and thus distance would be negative...